runtime/src/chpl-topo.c

```diff
@@ -0,0 +1,249 @@
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Compute node topology support
+//
+#include "chplrt.h"
+
+#include "chpl-align.h"
+#include "chplsys.h"
+#include "chpl-topo.h"
+#include "chpltypes.h"
+#include "error.h"
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+
+#ifdef CHPL_HAS_HWLOC
+//
+// We have hwloc, so provide a real implementation of the topology
+// interface.
+//
+#include "hwloc.h"
+
+#ifdef DEBUG
+// note: format arg 'f' must be a string constant
+#define _DBG_P(f, ...)                                                  \
+        do {                                                            \
+          printf("%s:%d" f "\n", __FILE__, __LINE__, ## __VA_ARGS__);   \
+        } while (0)
+#else
+#define _DBG_P(f, ...)
+#endif
+
+static hwloc_topology_t topology;
+
+static const struct hwloc_topology_support* topoSupport;
+
+static int topoDepth;
+
+static int numaLevel;
+static int numNumaDomains;
+
+
+static void chpl_topo_setMemLocalityByPages(unsigned char*, size_t,
+                                            hwloc_nodeset_t);
+static void report_error(const char*, int);
+
+
+void chpl_topo_init(void) {
+  //
+  // Allocate and initialize topology object.
+  //
+  if (hwloc_topology_init(&topology)) {
+    report_error("hwloc_topology_init()", errno);
+  }
+
+  //
+  // Limit discovery.
+  //
+  if (hwloc_topology_set_flags(topology, HWLOC_TOPOLOGY_FLAG_IS_THISSYSTEM)) {
+    report_error("hwloc_topology_set_flags()", errno);
+  }
+
+  //
+  // Perform the topology detection.
+  //
+  if (hwloc_topology_load(topology)) {
+    report_error("hwloc_topology_load()", errno);
+  }
+
+  //
+  // What is supported?
+  //
+  topoSupport = hwloc_topology_get_support(topology);
+
+  //
+  // We need depth information.
+  //
+  topoDepth = hwloc_topology_get_depth(topology);
+
+  //
+  // How many NUMA domains do we have?
+  //
+  {
+    int level;
+
+    //
+    // Note: if there are multiple levels with NUMA nodes, this just finds
+    // the uppermost.
+    //
+    for (level = 0, numaLevel = -1;
+         level < topoDepth && numaLevel == -1;
+         level++) {
+      if (hwloc_get_depth_type(topology, level) == HWLOC_OBJ_NUMANODE) {
+        numaLevel = level;
+      }
+    }
+  }
+
+  numNumaDomains = hwloc_get_nbobjs_by_depth(topology, numaLevel);
```

Comment:
Just a heads-up, this will give you the total number of NUMA nodes, including both compute nodes and memory-only nodes.  That's correct, just wanted to make sure you knew that.  It means that on KNL, it can return twice the number of compute nodes, so it can be different from chpl_task_getNumSublocales().